File: iterableapi/src/main/java/com/iterable/iterableapi/IterableActivityMonitor.java
package com.iterable.iterableapi;

import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.iterable.iterableapi.util.BrandUtils;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class IterableActivityMonitor {

    private static boolean initialized = false;
    private final Handler handler = new Handler(Looper.getMainLooper());
    private WeakReference<Activity> currentActivity;
    private int numStartedActivities = 0;
    private boolean inForeground = false;
    private List<WeakReference<AppStateCallback>> callbacks = new ArrayList<>();
    private Runnable backgroundTransitionRunnable = new Runnable() {
        @Override
        public void run() {
            inForeground = false;
            for (WeakReference<AppStateCallback> callback : callbacks) {
                if (callback.get() != null) {
                    callback.get().onSwitchToBackground();
                }
            }
        }
    };
    private static final int BACKGROUND_DELAY_MS = 1000;
    static IterableActivityMonitor instance = new IterableActivityMonitor();

    @NonNull
    public static IterableActivityMonitor getInstance() {
        return instance;
    }

    private Application.ActivityLifecycleCallbacks lifecycleCallbacks = new Application.ActivityLifecycleCallbacks() {
        @Override
        public void onActivityCreated(Activity activity, Bundle savedInstanceState) {

        }

        @Override
        public void onActivityStarted(Activity activity) {
            handler.removeCallbacks(backgroundTransitionRunnable);
            numStartedActivities++;
        }

        @Override
        public void onActivityResumed(Activity activity) {
            currentActivity = new WeakReference<>(activity);

            if (!inForeground || BrandUtils.isFireTV(activity.getPackageManager())) {
                inForeground = true;
                for (WeakReference<AppStateCallback> callback : callbacks) {
                    if (callback.get() != null) {
                        callback.get().onSwitchToForeground();
                    }
                }
            }
        }

        @Override
        public void onActivityPaused(Activity activity) {
            if (getCurrentActivity() == activity) {
                currentActivity = null;
            }
        }

        @Override
        public void onActivityStopped(Activity activity) {
            if (numStartedActivities > 0) {
                numStartedActivities--;
            }

            if (numStartedActivities == 0 && inForeground) {
                handler.postDelayed(backgroundTransitionRunnable, BACKGROUND_DELAY_MS);
            }
        }

        @Override
        public void onActivitySaveInstanceState(Activity activity, Bundle outState) {

        }

        @Override
        public void onActivityDestroyed(Activity activity) {

        }
    };

    public void registerLifecycleCallbacks(@NonNull Context context) {
        if (!initialized) {
            initialized = true;
            ((Application) context.getApplicationContext()).registerActivityLifecycleCallbacks(lifecycleCallbacks);
        }
    }

    public void unregisterLifecycleCallbacks(@NonNull Context context) {
        if (initialized) {
            initialized = false;
            ((Application) context.getApplicationContext()).unregisterActivityLifecycleCallbacks(lifecycleCallbacks);
        }
    }

    @Nullable
    public Activity getCurrentActivity() {
        return currentActivity != null ? currentActivity.get() : null;
    }

    public boolean isInForeground() {
        return getCurrentActivity() != null;
    }

    public void addCallback(@NonNull AppStateCallback callback) {
        // Don't insert again if the same callback already exists
        for (WeakReference<AppStateCallback> existingCallback : callbacks) {
            if (existingCallback.get() == callback) {
                return;
            }
        }
        callbacks.add(new WeakReference<>(callback));
    }

    public void removeCallback(@NonNull AppStateCallback callback) {
        Iterator<WeakReference<AppStateCallback>> iterator = callbacks.iterator();
        while (iterator.hasNext()) {
            WeakReference<AppStateCallback> callbackRef = iterator.next();
            if (callbackRef.get() == callback) {
                iterator.remove();
            }
        }
    }

    public interface AppStateCallback {
        void onSwitchToForeground();
        void onSwitchToBackground();
    }

}

File: iterableapi/src/main/java/com/iterable/iterableapi/IterableActivityMonitor.kt
package com.iterable.iterableapi

import android.app.Activity
import android.app.Application
import android.content.Context
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import androidx.annotation.NonNull
import androidx.annotation.Nullable
import com.iterable.iterableapi.util.DeviceInfoUtils
import java.lang.ref.WeakReference
import java.util.concurrent.CopyOnWriteArrayList

class IterableActivityMonitor {
    private val handler = Handler(Looper.getMainLooper())
    private var currentActivity: WeakReference<Activity>? = null
    private var numStartedActivities = 0
    private var inForeground = false
    private val callbacks: MutableList<WeakReference<AppStateCallback>> = CopyOnWriteArrayList()
    
    private val backgroundTransitionRunnable = Runnable {
        inForeground = false
        for (callback in callbacks) {
            callback.get()?.onSwitchToBackground()
        }
    }

    private val lifecycleCallbacks = object : Application.ActivityLifecycleCallbacks {
        override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
            // Empty implementation
        }

        override fun onActivityStarted(activity: Activity) {
            handler.removeCallbacks(backgroundTransitionRunnable)
            numStartedActivities++
        }

        override fun onActivityResumed(activity: Activity) {
            currentActivity = WeakReference(activity)

            if (!inForeground || DeviceInfoUtils.isFireTV(activity.packageManager)) {
                inForeground = true
                for (callback in callbacks) {
                    callback.get()?.onSwitchToForeground()
                }
            }
        }

        override fun onActivityPaused(activity: Activity) {
            if (getCurrentActivity() == activity) {
                currentActivity = null
            }
        }

        override fun onActivityStopped(activity: Activity) {
            if (numStartedActivities > 0) {
                numStartedActivities--
            }

            if (numStartedActivities == 0 && inForeground) {
                handler.postDelayed(backgroundTransitionRunnable, BACKGROUND_DELAY_MS.toLong())
            }
        }

        override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {
            // Empty implementation
        }

        override fun onActivityDestroyed(activity: Activity) {
            // Empty implementation
        }
    }

    fun registerLifecycleCallbacks(@NonNull context: Context) {
        if (!initialized) {
            initialized = true
            (context.applicationContext as Application).registerActivityLifecycleCallbacks(lifecycleCallbacks)
        }
    }

    fun unregisterLifecycleCallbacks(@NonNull context: Context) {
        if (initialized) {
            initialized = false
            (context.applicationContext as Application).unregisterActivityLifecycleCallbacks(lifecycleCallbacks)
        }
    }

    @Nullable
    fun getCurrentActivity(): Activity? {
        return currentActivity?.get()
    }

    fun isInForeground(): Boolean {
        return getCurrentActivity() != null
    }

    fun addCallback(@NonNull callback: AppStateCallback) {
        // Don't insert again if the same callback already exists
        for (existingCallback in callbacks) {
            if (existingCallback.get() == callback) {
                return
            }
        }
        callbacks.add(WeakReference(callback))
    }

    fun removeCallback(@NonNull callback: AppStateCallback) {
        for (callbackRef in callbacks) {
            if (callbackRef.get() == callback) {
                callbacks.remove(callbackRef)
            }
        }
    }

    interface AppStateCallback {
        fun onSwitchToForeground()
        fun onSwitchToBackground()
    }

    companion object {
        private var initialized = false
        private const val BACKGROUND_DELAY_MS = 1000
        val instance = IterableActivityMonitor()

        @NonNull
        @JvmStatic
        fun getInstance(): IterableActivityMonitor {
            return instance
        }
    }
}